### PR TITLE
Add spill and spiller tests with duplicated sort keys

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Add spill and spiller tests with duplicated sort keys

### CODE REVIEW

File: velox/exec/tests/SpillTest.cpp [hunk 1]
```diff
@@ -42,12 +42,21 @@ class SpillTest : public testing::Test,
...
+      int numDuplicates,
       int64_t expectedNumSpilledFiles) {
     auto tempDirectory = exec::test::TempDirectoryPath::create();
     const std::string kSpillPath = tempDirectory->path + "/test";
     std::shared_ptr<FileSystem> fs;
     std::vector<std::string> spilledFiles;
     const int kNumRowsPerBatch = 100'000;
+    std::vector<std::vector<int64_t>> values(numBatches);
+    std::vector<int64_t> expectedValues(kNumRowsPerBatch * numBatches * 2);
+    for (int i = 0, value = -10; i < kNumRowsPerBatch * numBatches * 2; ++value) {
```
Comment (mbasmanova) [inline velox/exec/tests/SpillTest.cpp:54]: It is a bit hard for me to understand the dataset being created here. Would you explain what the data looks like?

File: velox/exec/tests/SpillTest.cpp [hunk 2]
```diff
@@ -42,12 +42,21 @@ class SpillTest : public testing::Test,
...
       int numBatches,
+      int numDuplicates,
       int64_t expectedNumSpilledFiles) {
     auto tempDirectory = exec::test::TempDirectoryPath::create();
     const std::string kSpillPath = tempDirectory->path + "/test";
     std::shared_ptr<FileSystem> fs;
     std::vector<std::string> spilledFiles;
     const int kNumRowsPerBatch = 100'000;
+    std::vector<std::vector<int64_t>> values(numBatches);
+    std::vector<int64_t> expectedValues(kNumRowsPerBatch * numBatches * 2);
```
Comment (mbasmanova) [inline velox/exec/tests/SpillTest.cpp:53]: Are we creating numBatches batches of kNumRowsPerBatch * 2 rows each? kNumRowsPerBatch suggests that it is the number of rows in the batch, but it looks like the actual number of rows is double of that.

File: velox/exec/tests/SpillTest.cpp [hunk 3]
```diff
@@ -153,11 +158,14 @@ class SpillTest : public testing::Test,
 TEST_F(SpillTest, spillState) {
   // Set the target file size to a large value to avoid new file creation
   // triggered by batch write.
-  spillStateTest(1'000'000'000, 2, 10, 2 * 10);
+  spillStateTest(1'000'000'000, 2, 10, 1, 2 * 10);
+  spillStateTest(1'000'000'000, 2, 10, 10, 2 * 10);
```
Comment (mbasmanova) [inline velox/exec/tests/SpillTest.cpp:162]: nit: Perhaps, add a comment clarifying which call tests with duplicates and which without.

File: velox/exec/tests/SpillTest.cpp [hunk 4]
```diff
@@ -153,11 +158,14 @@ class SpillTest : public testing::Test,
...
+
 }
 
 TEST_F(SpillTest, spillStateWithSmallTargetFileSize) {
   // Set the target file size to a small value to open a new file on each batch
   // write.
-  spillStateTest(1, 2, 10, 2 * 10 * 2);
+  spillStateTest(1, 2, 10, 1, 2 * 10 * 2);
```
Comment (mbasmanova) [inline velox/exec/tests/SpillTest.cpp:169]: ditto

File: velox/exec/tests/SpillerTest.cpp [hunk 5]
```diff
@@ -194,16 +194,12 @@ class SpillerTest : public exec::test::RowContainerTestBase,
           }
           const int partitionNum =
               hashBits_.partition(hashes[0], numPartitions_);
-          if (numRowsPerPartition[partitionNum] == 0) {
-            selectedRows.setValid(index, false);
-            continue;
+          for (int i = 0;
+               i < numDuplicates && --numRowsPerPartition[partitionNum] >= 0;
+               ++i) {
+            rowVector_->copy(batch.get(), numFilledRows++, index, 1);
```
Comment (mbasmanova) [inline velox/exec/tests/SpillerTest.cpp:200]: nit: Perhaps, add a comment to explain which rows are being copied to rowVector_. I'm not sure I understand this code.

File: velox/exec/tests/SpillerTest.cpp [hunk 6]
```diff
@@ -353,19 +349,22 @@ class SpillerTest : public exec::test::RowContainerTestBase,
 };
 
 TEST_P(SpillerTest, spilFew) {
-  testSpill(10);
+  testSpill(10, 1);
```
Comment (mbasmanova) [inline velox/exec/tests/SpillerTest.cpp:352]: nit: perhaps, clarify which call tests with duplicates and which without.

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Thank you for increasing test coverage. The test code is a bit difficult to understand. I wonder if there is a way to make it clearer.

File: velox/exec/tests/SpillTest.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/tests/SpillTest.cpp:53]: Yeah, it is a little bit confusing.

File: velox/exec/tests/SpillTest.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply velox/exec/tests/SpillTest.cpp]: Updated. Thanks for pointing out!

**Comment by xiaoxmeng** [comment]
> @xiaoxmeng Thank you for increasing test coverage. The test code is a bit difficult to understand. I wonder if there is a way to make it clearer.

Thanks for the quick review and helpful suggestions!

**Pull request closed without merging**